src/theory/quantifiers/query_generator.cpp

```diff
@@ -0,0 +1,401 @@
+  return true;
+}
+
+void QueryGenerator::checkQuery(Node qy, unsigned spIndex)
+{
+  // external query
+  if (options::sygusQueryGenDumpFiles())
+  {
+    // Print the query and the query + its model (commented) to queryN.smt2
+    std::vector<Node> pt;
+    d_sampler->getSamplePoint(spIndex, pt);
+    unsigned nvars = d_vars.size();
+    AlwaysAssert(pt.size() == d_vars.size());
+    std::stringstream fname;
+    fname << "query" << d_query_count << ".smt2";
+    std::ofstream fs(fname.str(), std::ofstream::out);
+    fs << "(set-logic ALL)" << std::endl;
+    for (unsigned i = 0; i < 2; i++)
+    {
+      for (unsigned j = 0; j < nvars; j++)
+      {
+        Node x = d_vars[j];
+        if (i == 0)
+        {
+          fs << "(declare-fun " << x << " () " << x.getType() << ")";
+        }
+        else
+        {
+          fs << ";(define-fun " << x << " () " << x.getType() << " " << pt[j]
+             << ")";
+        }
+        fs << std::endl;
+      }
+    }
+    fs << "(assert " << qy << ")" << std::endl;
+    fs << "(check-sat)" << std::endl;
+    fs.close();
+  }
+
+  if (options::sygusQueryGenCheck())
+  {
+    Trace("sygus-qgen-check") << "  query: check " << qy << "..." << std::endl;
+    NodeManager* nm = NodeManager::currentNM();
+    // make the satisfiability query
+    bool needExport = false;
+    ExprManagerMapCollection varMap;
+    ExprManager em(nm->getOptions());
+    std::unique_ptr<SmtEngine> queryChecker;
+    initializeChecker(queryChecker, em, varMap, qy, needExport);
+    Result r = queryChecker->checkSat();
+    Trace("sygus-qgen-check") << "  query: ...got : " << r << std::endl;
+    if (r.asSatisfiabilityResult().isSat() == Result::UNSAT)
+    {
+      std::stringstream ss;
+      ss << "--sygus-rr-query-gen detected unsoundness in CVC4 on input " << qy
+         << "!" << std::endl;
+      ss << "This query has a model : " << std::endl;
+      std::vector<Node> pt;
+      d_sampler->getSamplePoint(spIndex, pt);
+      Assert(pt.size() == d_vars.size());
+      for (unsigned i = 0, size = pt.size(); i < size; i++)
+      {
+        ss << "  " << d_vars[i] << " -> " << pt[i] << std::endl;
+      }
+      ss << "but CVC4 answered unsat!" << std::endl;
+      AlwaysAssert(false, ss.str().c_str());
+    }
+  }
+
+  d_query_count++;
+}
+
+void QueryGenerator::findQueries(
+    Node n,
+    std::vector<Node>& queries,
+    std::vector<std::vector<unsigned>>& queriesPtTrue)
+{
+  TypeNode tn = n.getType();
+  LazyTrie* lt = &d_qgt_trie[tn];
+  std::vector<unsigned> eqIndex[2];
+  Trace("sygus-qgen-debug") << "Compute queries for " << n << "...\n";
+
+  LazyTrieEvaluator* ev = d_sampler;
+  unsigned ntotal = d_sampler->getNumSamplePoints();
+  unsigned index = 0;
+  bool exact = true;
+  bool pushEq[2] = {false, false};
+  bool pre = true;
+  std::vector<LazyTrie*> visitTr;
+  std::vector<unsigned> currIndex;
+  std::vector<bool> currExact;
+  std::vector<bool> pushIndex[2];
```

Comment:
Suggestion: use a vector of pairs. We are always pushing/popping both vectors at the same time.

Reply:
The issue is I iterate `for( unsigned r=0; r<2; r++)` to access `pushIndex[r]` in a number of places. Is there a way to look up first/second via index `r=0,1` ?

Reply:
I don't think so because they can have different types. But my suggestion was actually to make it an `std::vector<std::pair>`, so instead of doing:
```
    for (unsigned r = 0; r < 2; r++)
    {
      pushEq[r] = pushIndex[r].back();
    }
```
You'd simply do `std::pair<bool, bool> pushEq = pushIndex.back();`. For inserting, you would do `pushIndex.push_back({false, false});`